hloc/extractors/disk.py

```diff
@@ -0,0 +1,109 @@
+        'model_name': 'depth-save.pth', # Name of the model's .pth save file
+        # we rescale image only in the preprocessing step.
+        # we only pad the image size to be the multiple of 16 in this model
+        # 'height': None, # rescaled height (px). If unspecified, image is not resized in height dimension
+        # 'width': None, # rescaled width (px). If unspecified, image is not resized in width dimension
+        'window':  5, # NMS window size
+        'n': None, # Maximum number of features to extract. If unspecified, the number is not limited
+        'desc-dim': 128, # descriptor dimension. Needs to match the checkpoint value
+        'mode': 'nms', # Whether to extract features using the non-maxima suppresion mode or through training-time grid sampling technique
+    }
+    required_inputs = ['image']
+
+    def _init(self, conf):
+        state_dict = torch.load(disk_path / conf['model_name'], map_location='cpu')
+
+        if 'extractor' in state_dict:
+            weights = state_dict['extractor']
+        elif 'disk' in state_dict:
+            weights = state_dict['disk']
+        else:
+            raise KeyError('Incompatible weight file!')
+        self.model = _DISK(window=8, desc_dim=conf['desc-dim'])
+        self.model.load_state_dict(weights)
+        self.model.to(torch.device('cuda' if torch.cuda.is_available() else 'cpu'))
+        if conf['mode'] == 'nms':
+            self.extract = partial(
+                self.model.features,
+                kind='nms',
+                window_size=conf['window'],
+                cutoff=0.,
+                n=conf['n']
+            )
+        elif conf['mode'] == 'rng':
+            self.extract = partial(self.model.features, kind='rng')
+        else:
+            raise KeyError('mode must be either nms or rng!')
+
+    def _forward(self, data):
+        img = data['image'][0]
+        assert len(img.shape) == 3 and img.shape[0] == 3
+        # pad img so that its height and width be the multiple of 16 as required by the original repo
+        orig_h, orig_w = img.shape[1:]
+        new_h = ((orig_h-1)//16 + 1) * 16
+        new_w = ((orig_w-1)//16 + 1) * 16
+        y_pad = new_h - orig_h
+        x_pad = new_w - orig_w
+
+        img = F.pad(img, (0, x_pad, 0, y_pad))
+        assert img.shape[1] == new_h and img.shape[2] == new_w, "Wrong Padding!"
+
+        with torch.no_grad():
+            try:
+                batched_features = self.extract(img[None])  # add batch dimension
+            except RuntimeError as e:
+                if 'U-Net failed' in str(e):
+                    msg = ('Please use input size which is multiple of 16 (or '
+                           'adjust the --height and --width flags to let this '
+                           'script rescale it automatically). This is because '
+                           'we internally use a U-Net with 4 downsampling '
+                           'steps, each by a factor of 2, therefore 2^4=16.')
```

Comment:
Maybe the error message should be changed now

Reply:
I would like to remove this try-except block as we explicitly pad the image to be the multiple of 16.